### PR TITLE
End-To-End Testing Framework

### PR DESCRIPTION
**Pull request by Arvinderpal**
As mentioned in https://github.com/cloudnativelabs/kube-router/issues/624, kube-router needs a e2e testing framework. It is particularly important for the ipv6 and dual-stack work, since they will require substantial changes to kube-router. 

In summary, this framework 
* makes use of ginkgo and gomega
* allows for tests can be run against an existing local vagrant cluster
* includes some initial basic tests for router and  service-proxy functions.
* is flexible enough to allow additional provisioners to be added (e.g. digital ocean)

In order to test the framework, you will have to first provision a k8 cluster. Certain cluster variables must be fixed; for example, k8 master must be called "k8s1" and worker "k8s2". Alternatively, you can modify these values in `helpers/cons.go` 

I have tested the framework against the following vagrant based cluster: https://github.com/Nordix/k8s-ipv6

Please refer to the README.md file in the `test` directory for details. 

As a follow up PR, I will add auto provisioning of vagrant based clusters.

### CODE REVIEW

**Comment by Arvinderpal** [comment]
@murali-reddy @andrewsykim 
Hi guys, how should be go about reviewing this? I realize it's a large PR (though vendoring makes up the vast majority of the bulk), so I want to do whatever possible to make life easier for the reviewers. I would be willing to do an online sync and code walk as well if that would help. Please advise.

File: test/config/config.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	TestScope        string
+	SkipLogGathering bool
+}
+
+// KuberouterTestConfig holds the global configuration of commandline flags
+// in the ginkgo-based testing environment.
+var KuberouterTestConfig = KuberouterTestConfigType{}
+
+// ParseFlags parses commandline flags relevant to testing.
+func (c *KuberouterTestConfigType) ParseFlags() {
```
Comment (andrewsykim) [inline test/config/config.go]: Can we specify defaults in brackets for each flag description?

File: test/helpers/cons.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 Authors of Cilium
```
Comment (andrewsykim) [inline test/helpers/cons.go]: Prefer to call this file `constants.go`

File: test/helpers/gobgp.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// ExpectGobgpNeighEstabl asserts via gobgp that neighbor is in Established state
+func (s *SSHMeta) ExpectGobgpNeighEstabl() {
+	err := s.WaitUntilNeighborEstabl(100)
+	ExpectWithOffset(1, err).To(BeNil(), "gobgp neighbor not in established state")
+}
+
+// ExecGobgpNeighCheck checks neighbor node via gobgp
+func (s *SSHMeta) ExecGobgpNeighCheck() *CmdRes {
+	var command string
+	command = fmt.Sprintf("gobgp neighbor")
```
Comment (andrewsykim) [inline test/helpers/gobgp.go]: no need for the extra var here if it's just used here.

```
return s.Exec("gobgp neighbor")
```

File: test/helpers/kubectl.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+	node.logger = log
+
+	return &Kubectl{
+		SSHMeta: node,
+	}
+}
+
+// ExpectAllPodsTerminated is a wrapper around helpers/WaitCleanAllTerminatingPods.
+// It asserts that the error returned by that function is nil.
+func (kub *Kubectl) ExpectAllPodsTerminated() {
```
Comment (andrewsykim) [inline test/helpers/kubectl.go]: prefer `k` or `kctl` rather than `kub`.

File: test/helpers/kubectl.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+	err := kub.WaitCleanAllTerminatingPods(HelperTimeout)
+	ExpectWithOffset(1, err).To(BeNil(), "terminating containers are not deleted after timeout")
+}
+
+// WaitCleanAllTerminatingPods waits until all nodes that are in `Terminating`
+// state are deleted correctly in the platform. In case of excedding the
+// given timeout (in seconds) it returns an error
+func (kub *Kubectl) WaitCleanAllTerminatingPods(timeout int64) error {
+	body := func() bool {
+		res := kub.Exec(fmt.Sprintf(
```
Comment (andrewsykim) [inline test/helpers/kubectl.go]: Any reason for not using client go here? That would give us more flexibility on getting/creating/deleting pods and would be a bit cleaner IMO

Comment (andrewsykim) [reply test/helpers/kubectl.go]: Guessing this is just from copying it over from Cilium?

File: test/router/pod2pod.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
+package RouterTest
```
Comment (andrewsykim) [inline test/router/pod2pod.go]: prefer:
```go
package router_tests
```
or even just
```
package router
```

File: test/runtime/log.go [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package RuntimeTest
```
Comment (andrewsykim) [inline test/runtime/log.go]: same here `package runtime ` or `package runtime_test`

File: test/service-proxy/connectivity.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
+package ServiceProxyTest
```
Comment (andrewsykim) [inline test/service-proxy/connectivity.go]: `package service_proxy_test` or `package service_proxy`

File: test/test_suite_test.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
+package kuberouterTest
```
Comment (andrewsykim) [inline test/test_suite_test.go]: same

File: test/helpers/kubectl.go (hunk 5 above)
Comment (Arvinderpal) [reply test/helpers/kubectl.go]: Yeah, I pretty much stuck to using the cilium approach. Being able to issues other commands beyond kubectl is also a nice benefit of this approach. We could always add "k8s.io/client-go" in the future if we have the need to do more complex stuff...

**Comment by Arvinderpal** [comment]
> Are you able to provide some example output of logs/results from when you run it in your environment?
@andrewsykim I will update the README with some example output. I can also slack you the files if desired. 
Of course, the best way would be to create a local vagrant cluster of your own and run the tests against it. There are a few constraints to keep in mind. I can help you with those. I will add auto provisioning of VMs in the next PR as this PR is already pretty large.

**Comment by Arvinderpal** [comment]
Hi guys, you can create a vagrant based ipv6 cluster here: https://github.com/Nordix/k8s-ipv6
After doing so, you should be able to run the e2e tests against that cluster. 
Please let me know if you have any issues.

**Comment by andrewsykim** [comment]
@Arvinderpal sorry I haven't had a chance to review this yet, will get to it soon. Thanks for your patience!

**Comment by asteven** [comment]
@murali-reddy could you please have a look at this? Despite that I'd love to have ipv6 support in kube-router it would be very, very useful to have better testing possibilities in place. There are some interesting pull-requests piling up and having a way to properly test that they don't have any side effects would make it much easier to review and possibly merge them with more confidence.

**Comment by asteven** [comment]
Hi @Arvinderpal,

Thanks for working on this. Much appreciated! You totally picked up something that kube-router needs badly.

I've been looking at the PR and noticed that your e2e framework is pretty much self contained. It does not really change kube-router's existing codebase, the coupling is pretty loose.

Wouldn't it be better to develop and maintain it in a dedicated repo?

**Comment by Arvinderpal** [comment]
Hi @asteven, sorry about the delay. Yes, @murali-reddy created the following repo and I will be moving this PR to that repo. 
https://github.com/cloudnativelabs/kube-router-e2e

cheers,

**Comment by Arvinderpal** [comment]
Moved to: 
https://github.com/cloudnativelabs/kube-router-e2e

**Comment by asteven** [comment]
@Arvinderpal Thanks for your contribution!

**Pull request closed without merging**